src/Nethermind/Nethermind.JsonRpc/JsonRpcUrlCollection.cs

```diff
@@ -0,0 +1,109 @@
+
+            if (_jsonRpcConfig.Enabled)
+                BuildUrls(includeWebSockets);
+        }
+
+        public string[] Urls => Values.Select(x => x.ToString()).ToArray();
+
+        private void BuildUrls(bool includeWebSockets)
+        {
+            JsonRpcUrl defaultUrl = new JsonRpcUrl(Uri.UriSchemeHttp, _jsonRpcConfig.Host, _jsonRpcConfig.Port, RpcEndpoint.Http, _jsonRpcConfig.EnabledModules);
+            string environmentVariableUrl = Environment.GetEnvironmentVariable(NethermindUrlVariable);
+            if (!string.IsNullOrWhiteSpace(environmentVariableUrl))
+            {
+                if (Uri.TryCreate(environmentVariableUrl, UriKind.Absolute, out Uri? uri))
+                {
+                    defaultUrl.Scheme = uri.Scheme;
+                    defaultUrl.Host = uri.Host;
+                    defaultUrl.Port = !uri.IsDefaultPort ? uri.Port : defaultUrl.Port;
+                }
+                else if (_logger.IsWarn)
+                    _logger.Warn($"Environment variable '{NethermindUrlVariable}' value '{environmentVariableUrl}' is not valid JSON RPC URL, using default url : '{defaultUrl}'");
+            }
+            Add(defaultUrl.Port, defaultUrl);
+
+            if (includeWebSockets)
+            {
+                if (_jsonRpcConfig.WebSocketsPort != _jsonRpcConfig.Port)
+                {
+                    JsonRpcUrl defaultWebSocketUrl = defaultUrl.Clone() as JsonRpcUrl;
+                    defaultWebSocketUrl.Port = _jsonRpcConfig.WebSocketsPort;
+                    defaultWebSocketUrl.RpcEndpoint = RpcEndpoint.Ws;
+                    Add(defaultWebSocketUrl.Port, defaultWebSocketUrl);
+                }
+                else
+                    defaultUrl.RpcEndpoint |= RpcEndpoint.Ws;
+            }
+
+            foreach (string additionalRpcUrl in _jsonRpcConfig.AdditionalRPCUrls)
+            {
+                try
+                {
+                    JsonRpcUrl url = JsonRpcUrl.Parse(additionalRpcUrl);
+                    if (!includeWebSockets && url.RpcEndpoint.HasFlag(RpcEndpoint.Ws))
+                    {
+                        url.RpcEndpoint &= ~RpcEndpoint.Ws;
+                        if (url.RpcEndpoint == RpcEndpoint.None)
+                        {
+                            if (_logger.IsInfo)
+                                _logger.Info($"Additional JSON RPC URL '{url}' has web socket endpoint type and web sockets are not enabled; skipping...");
+                            continue;
+                        }
+                    }
+
+                    if (ContainsKey(url.Port))
+                    {
+                        if (_logger.IsWarn)
+                        { _logger.Warn($"Additional JSON RPC URL '{url}' wants port {url.Port}, but port already in use; skipping..."); }
+                        continue;
+                    }
+
+                    Add(url.Port, url);
+                }
+                catch (Exception)
```

Comment:
Shouldn't we handle only FormatException here?

Reply:
It needs to catch any exception that JsonRpcUrl.Parse() can throw to allow it to recover, log the warning that the url parse failed, then skip it, and continue parsing the next url. Maybe the warning should be reworded from "not formatted correctly" to "could not be parsed"

Reply:
Refactored to make JsonRpcUrl.Parse() thrown exceptions consistent by using FormatException; Also passing on relevant and helpful format exception messages in the log.